lib/ssl-test.rb

```diff
@@ -56,4 +62,87 @@ def self.matching_domains domains, hostname
     domains.map {|s| Regexp.new("\A#{Regexp.escape(s).gsub('\*', '[^.]+')}\z") }
       .select {|domain| domain.match?(hostname) }
   end
+
+  # https://docs.ruby-lang.org/en/2.2.0/OpenSSL/OCSP.html
+  def self.test_ocsp_revocation cert, chain, open_timeout: 5, read_timeout: 5, redirection_limit: 5
+    issuer = chain.last
+
+    digest = OpenSSL::Digest::SHA1.new
+    certificate_id = OpenSSL::OCSP::CertificateId.new(cert, issuer, digest)
+
+    request = OpenSSL::OCSP::Request.new
+    request.add_certid certificate_id
+    request.add_nonce
+
+    authority_info_access = cert.extensions.find do |extension|
+      extension.oid == "authorityInfoAccess"
+    end
+
+    descriptions = authority_info_access.value.split("\n")
+    ocsp = descriptions.find do |description|
+      description.start_with?("OCSP")
+    end
+
+    ocsp_uri = URI(ocsp[/URI:(.*)/, 1])
+    # "Note that we only handle HTTP requests and don't handle any redirects in this example"
+    http_response = follow_ocsp_redirects(ocsp_uri, request.to_der, open_timeout: open_timeout, read_timeout: read_timeout, redirection_limit: redirection_limit)
+    return [false, nil, nil] unless http_response
+
+    response = OpenSSL::OCSP::Response.new http_response.body
+    basic_response = response.basic
+
```

Comment:
At this point we're supposed to check that the response has a valid signature.

To do this, we should use [`OpenSSL::X509::Store`](https://ruby-doc.org/stdlib-2.4.0/libdoc/openssl/rdoc/OpenSSL/X509/Store.html) which would give something along the lines of:
```ruby
store = OpenSSL::X509::Store.new
store.set_default_paths
response.verify([], store) # This needs to be true
```

[`OpenSSL::X509::Store#set_default_paths`](https://ruby-doc.org/stdlib-2.4.0/libdoc/openssl/rdoc/OpenSSL/X509/Store.html#method-i-set_default_paths) would look up CA certificates from the system default certificate store, but if they're not properly set up, it'd say that the response can't be trusted.

The doc speaks about this case and says that we "can obtain a set extracted from Mozilla CA certificate store by cURL maintainers" and then use [`OpenSSL::X509::Store#add_file`](https://ruby-doc.org/stdlib-2.4.0/libdoc/openssl/rdoc/OpenSSL/X509/Store.html#method-i-add_file).

I skipped all this part for convenience, but it makes it unsafe because the OCSP responder could be impersonated.

Reply:
Good point, we should do the validation with the default store IMO. It's ok on most modern linux distributions and it'll be the same used for cert validation anyway so it makes sense.